File: 0146-lru-cache/0146-lru-cache.java
class LRUCache<K,V> {

    Node<K,V> dummyHead = new Node<>(null, null);
    Node<K,V> dummyTail = new Node<>(null, null);
    int capacity = 0;
    Map<K, Node<K,V>> map;

    public LRUCache(int capacity) {
        dummyHead.next = dummyTail;
        dummyTail.prev = dummyHead;
        this.capacity = capacity;
        map = new HashMap<>();
    }
    
    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }

        Node<K,V> getNode = map.get(key);
        removeNode(getNode);
        addNode(getNode);

        return (Integer) getNode.value;
        
    }
    
    public void put(K key, V value) {

        if (map.containsKey(key)) {
            Node<K,V> currNode = map.get(key);
            currNode.value = value;
            map.put(key, currNode);

            removeNode(currNode);
            addNode(currNode);
            return;
        }

        if (map.size() == capacity) {
            // evict
            Node<K,V> evict = dummyTail.prev;
            map.remove(evict.key);
            removeNode(evict);
        }

        Node<K, V> node = new Node<K, V>(key, value);
        map.put(key, node);
        addNode(node);

    }

    public void addNode(Node<K,V> newNode) {
        // add front
        Node<K,V> currFront = dummyHead.next;

        newNode.prev = dummyHead;
        dummyHead.next = newNode;

        newNode.next = currFront;
        currFront.prev = newNode;

    }

    public void removeNode(Node<K,V> node) {
        Node<K,V> prevNode = node.prev;
        Node<K,V> nextNode = node.next;

        prevNode.next = nextNode;
        nextNode.prev = prevNode;
    }

    class Node<K, V> {
        K key;
        V value;
        Node<K,V> next;
        Node<K,V> prev;

        public Node(K key, V value) {
            this.key = key;
            this.value = value;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.head = Node(-1, -1)
        self.tail = Node(-1,-1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.map = {}
    

    def get(self, key: int) -> int:
        if key in self.map:
            node = self.map[key]
            self.deleteNode(node)
            self.addNode(node)
            return node.value

        return -1    
        

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            node = self.map[key]
            new_node = Node(key, value)
            self.deleteNode(node)
            self.addNode(new_node)
            self.map[key] = new_node   
        else:
            if len(self.map) == self.capacity:
                evict = self.tail.prev
                self.deleteNode(evict)
                del self.map[evict.key]
                
            new_node = Node(key, value)
            self.addNode(new_node)
            self.map[new_node.key] = new_node    


    # add always to the front
    def addNode(self, node):
        curr_head_next = self.head.next
        self.head.next = node
        node.prev = self.head
        node.next = curr_head_next
        curr_head_next.prev = node

    # delete happens from the last
    def deleteNode(self, node):
        prev_node = node.prev
        next_node = node.next
        prev_node.next = next_node
        next_node.prev = prev_node
        

class Node:

    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None
    

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)